pkg/actuators/machine/instances.go

```diff
@@ -363,68 +376,142 @@ func getImageID(machine runtimeclient.ObjectKey, machineProviderConfig *alibabac
 	return image.ImageId, nil
 }
 
-func getSecurityGroupID(machine runtimeclient.ObjectKey, machineProviderConfig *alibabacloudproviderv1.AlibabaCloudMachineProviderConfig, client alibabacloudClient.Client) (string, error) {
+func getSecurityGroupIDs(machine runtimeclient.ObjectKey, machineProviderConfig *alibabacloudproviderv1.AlibabaCloudMachineProviderConfig, client alibabacloudClient.Client) (*[]string, error) {
 	klog.Infof("%s validate security group in region %s", machineProviderConfig.SecurityGroupID, machineProviderConfig.RegionID)
+	var securityGroupIDs []string
+
+	// If SecurityGroupID is assigned, use it directly
+	if machineProviderConfig.SecurityGroupID != "" {
+		securityGroupIDs = append(securityGroupIDs, machineProviderConfig.SecurityGroupID)
+	} else {
+		// Otherwise, the query securityGroupIDs by the tags
+		for _, sg := range machineProviderConfig.SecurityGroups {
+			if sg.ID != "" {
+				securityGroupIDs = append(securityGroupIDs, sg.ID)
+			} else {
+				if sg.Tags != nil {
+					request := ecs.CreateDescribeSecurityGroupsRequest()
+					if machineProviderConfig.VpcID != "" {
+						request.VpcId = machineProviderConfig.VpcID
+					}
+					if machineProviderConfig.ResourceGroupID != "" {
```

Comment:
When `ResourceGroupID` is specified I am unable to query for the correct security group. I would remove resource group id.
```
aliyun ecs DescribeSecurityGroups --RegionId us-east-1 \
 --Tag.0.Key owned --Tag.0.Value "kubernetes.io/cluster/test-2mbh7" \
 --Tag.1.Key Name --Tag.1.Value "test-2mbh7-sg-worker" \
 --Tag.2.Key OCP --Tag.2.Value "ISV Integration"
{
	"PageNumber": 1,
	"PageSize": 10,
	"RegionId": "us-east-1",
	"RequestId": "BCD4DF49-FEED-57F6-A3AE-D670FB876026",
	"SecurityGroups": {
		"SecurityGroup": [
			{

				"CreationTime": "2021-11-01T17:11:52Z",
				"Description": "Created By OpenShift Installer",
				"ResourceGroupId": "",
				"SecurityGroupId": "sg-0xi9838n99r7i86s6rri",
				"SecurityGroupName": "test-2mbh7-sg-worker",
				"SecurityGroupType": "normal",
				"ServiceManaged": false,
				"Tags": {
					"Tag": [
						{
							"TagKey": "kubernetes.io/cluster/test-2mbh7",
							"TagValue": "owned"
						},
						{
							"TagKey": "Name",
							"TagValue": "test-2mbh7-sg-worker"
						},
						{
							"TagKey": "OCP",
							"TagValue": "ISV Integration"
						}
					]
				},
				"VpcId": "vpc-0xiosfgntngzl5ali1wj5"
...
```
Notice the `resourceGroupID` is not set.

Reply:
I think if you don't assign an instance(ig. ECS Instance Or SecurityGroup ,etc) to a resource group. You don't need to set the resourceGroupID. If you set the resourceGroupID. The API simply returns the resources under a given resource group.

Reply:
@kwoodson I think I can explain the problem, it should be that the resource_group is not specified [here](https://github.com/openshift/installer/blob/7a9bbc704b601a2c4af8821bfa3daeb8e9c64323/data/data/alibabacloud/cluster/vpc/sg-worker.tf#L1)
`resource_group_id` parameter, I have fixed it and submitted it in #5348